### PR TITLE
Support Dhall standard version 16.0.0

### PR DESCRIPTION
**Pull request by Gabriel439**

### CODE REVIEW

File: dhall/src/Dhall/Parser/Expression.hs [hunk 1]
```diff
@@ -255,14 +260,59 @@ parsers embedded = Parsers {..}
...
+
+                        return (\e -> With e keys value) )
+
+                    return (foldl (\e f -> f e) a0 bs)
+
+            alternative4A <|> alternative4B <|> alternative4C <|> pure a
+
+    -- The firstApplicationExpression argument is necessary in order to
+    -- left-factor the parsers for function types and @with@ expressions to
+    -- minimize backtracking
```
Comment (sjakobi) [inline dhall/src/Dhall/Parser/Expression.hs:302]: Could you possibly expand on this a bit? Currently I'm quite stumped. (Admittedly I never really had a good grasp on this part of the parser.)

Comment (Gabriel439) [reply dhall/src/Dhall/Parser/Expression.hs:302]: @sjakobi: This is to prevent the same parsing performance regression as in: https://github.com/dhall-lang/dhall-haskell/issues/108#issuecomment-425152274

Without this complicated logic we have to backtrack when parsing the function's input type, which kills performance.

The easiest way to explain the current algorithm is by incrementally fixing simple but incorrect solutions.

## Approach 1

* Attempt to parse a function's input type
* if that fails, then backtrack and try to parse a `with` expression instead

This is correct, but slow for the same reasons as #108 because of the backtracking

## Approach 2

Try to combine ("left factor') the parsers for a function type and a `with` expression to avoid backtracking.

The general idea behind left-factoring is that if you pretend that `<|>` is `+` and pretend that `*>` is `×`, then you can make a parser more efficient by turning code like this:

```
(a × b) + (a × c)

Analogous to: (a *> b) <|> (a *> c)
```

... into code like this:

```
a × (b + c)

Analogous to: a *> (b <|> c)
```

It's called "left factoring" the parser because it's like factoring out a shared multiplier on the left according to the rules of arithmetic.

In this case, we want to factor out the parser for a function's input type to share the same work as the parser for the left argument of a `with` expression.  This shared parser corresponds to the `a` in the above arithmetic expression.  Then the `b` would be the parser for the rest of the function type and the `c` would be the parser for the rest of the `with` expression.

So the left-factored algorithm is:

* attempt to parse an operator expression, which would be either (A) the input to a function type, (B) the left argument to a `with` expression, or (C) just an ordinary operator expressoin
* attempt to parse the remainder of a function type
* if that fails, then attempt to parse the remainder of a `with` expression
* if that fails, then treat it as an ordinary operator expression

This fixes the performance issue because we don't backtrack while parsing the initial operator expression, but it is incorrect, because the left argument to a `with` expression cannot be an operator expression.  The left argument of a `with` is restricted to an import expression.

## Approach 3

We need a way to try parsing the prefix as both an import expression (as a potential candidate for the left argument to `with`) and an operator expression (as a potential candidate for a function's input type), without backtracking.  In order do to this, we need to:

* First attempt to parse an import expression, which might be the left argument to a `with` expression
* Now parse the remainder of the operator expression, reusing what we've parsed so far from import expression we just parsed

  This is the hard part which introduces most of the complexity because we need a way to subtract out the import expression that we parsed so far so that we don't attempt to parse it again when parsing the remainder of the operator expression.

* Attempt to parse the remainder of a function type
* If that fails, then attempt to parse the remainder of a `with` expression, but only if the operator expression was the same as the import expression
* If that fails, then it is an ordinary operator expression

The second to last step is why the code checks to see if the operator expression was not actually an operator:

https://github.com/dhall-lang/dhall-haskell/pull/1770/files#diff-281e7ffb783d37c08218e547d14115b2R264-R279

I could have also checked that the import expression and the operator expression were one and the same by writing:

```haskell
guard (shallowDenote a == a0)
```

... but that would incur a new `Eq` constraint on the `Parser` which I was trying to avoid.

**TL;DR:** This sort of left-factoring complication will go away once we switch to an LR parser generator, but we can't do that until we remove the deprecated `exprA` utility from the API.

**Comment by sjakobi** [comment]
BTW I thought that the PR title was slightly confusing. I think you mostly used the "Version X → Y" pattern in this repository with reference to the `dhall` package version. Here it's a reference to the version of the Dhall standard.

**Comment by Gabriel439** [comment]
@sjakobi: Yeah, I mistakenly used the wrong title here.  Let me fix it.

**Comment by Gabriel439** [comment]
Actually, I think I will rename this one more time to `Support Dhall standard version 1.16.0`. That way we can have a separate pull request for actually bumping this package's version, including the `CHANGELOG` changes.

**Comment by amarrella** [comment]
@Gabriel439 isn't the standard version 16.0.0 ?

**Comment by Gabriel439** [comment]
@amarrella: Yes, it is :slightly_smiling_face: 

Fixed
